man/girl.1
.\" Man Page for gPhoto by scott... :)
.\" groff -man -Tascii foo.1
.\"
.TH GIRL 1 "NOVEMBER 2014" GNOME "User Manuals"
.SH NAME
girl \- GNOME Internet Radio Locator
.SH SYNOPSIS
.B girl
.br
.SH DESCRIPTION
.B girl,
the GNOME Internet Radio Locator program, allows users to easily find live radio programs on radio broadcasters on the Internet.
.SH MOTIVATION
The motivation behind this program is to support the user with easy access to independent, free Internet radio stations around the world. Enjoy Internet Radio.
.SH PLATFORM
.B girl
is developed on the new GNOME 2.0 platform and requires at least one audio helper such as totem to be installed.
.SH USAGE
.IP girl
.I "\--help"
display help screen
.SH LOCAL STATIONS FILE
.I $HOME/.girl/stations.xml
.RS
Radio station XML file
.SH STATION CONFIGURATION FILE
.I $HOME/.gnome2/girl
.RS
Default radio station configuration
.SH BUGS
http://bugzilla.gnome.org/browse.cgi?product=girl
.SH UPDATES
https://git.gnome.org/girl
.SH AUTHORS
.B 
Ole Aamot <ole@src.gnome.org>
.SH "SEE ALSO"
.BR totem


NEWS
2014-12-14  Ole Aamot  <oka@oka.no>

	* NEWS: girl version 0.7.3 release
	* src/stations.xml: Add CJSW (cjsw.com)
	* src/stations.xml: Add Caper Radio (caperradio.ca)

2014-12-14  Ole Aamot  <oka@oka.no>

	* NEWS: girl version 0.7.2 "Don't shoot" release
	* po/pl.po: Updated Polish translation by Piotr Drąg
	* src/stations.xml: Add WAMU (wamu.org)
	* src/stations.xml: Add WXOU (wxou.org)
	
2014-12-13  Ole Aamot  <oka@oka.no>

	* AUTHORS: Add Muhammet Kara (Turkish translation)
	* THANKS: Muhammet Kara for Turkish translation
	* girl version 0.7.1 release.
	* data/girl.appdata.xml.in: Remove old reference to "new GNOME 2.0".
	* src/girl-0.7.dtd: Add girl 0.7 DTD with station language attribute
	* src/girl-gui.c: Add Streams and About Stream buttons
	* src/girl-station.c: Add new function girl_stream_record
	* src/girl-streams.c: New file
	* src/girl-streams.h: New file
	* src/stations.xml: Add 95bFM (95bfm.com)
	* src/stations.xml: Add KDUP (kdup.up.edu)
	* src/streams.xml: Add streams file

2014-12-12  Ole Aamot  <oka@oka.no>

	* girl version 0.7.0 release.
	* AUTHORS: Add Piotr Drąg (Polish translation)
	* NEWS: Release girl version 0.7.0
	* THANKS: Piotr Drąg for minor typo fix and Polish translation
	* girl.spec.in: Update RPM spec file for Fedora 21
	* po/pl.po: Polish translation by Piotr Drąg
	* src/girl-gui.c: Add GtkWidget progress
	* src/girl-runners.c: Modify mutex functions for future usage.
	* src/girl-runners.h: Modify mutex functions for future usage.
	* src/girl-station.c: Add asynchronous child process function.
	* src/girl.h: Rename DEBUG macro GIRL_DEBUG, add progress bar.
	* src/stations.xml: Add 2NURFM (broadcaster1.newcastle.edu.au)
	* src/stations.xml: Add Radio Adelaide (radio.adelaide.edu.au)

2014-12-10  Ole Aamot  <oka@oka.no>

	* girl version 0.6.283185307 release in honor of Malala
	* src/girl-runners.c: Add file for the threads implementation
	* src/girl-runners.h: Add file for the threads implementation

2014-12-07  Ole Aamot  <oka@oka.no>

	* girl version 0.6.2 release.
	* src/stations.xml: Add Radio R (radior.cz)
	* src/stations.xml: Add Radio Radius (radioradius.ch)

2014-12-07  Ole Aamot  <oka@linux.oka.no>

	* girl version 0.6.1 release.
	* THANKS: Add Ettore Perazzoli and Bernd Homuth.
	* configure.ac: Check for GIRL_HELPER_PLAYER (totem)
	* configure.ac: Check for GIRL_HELPER_RECORD (streamripper)
	* configure.ac: Check for and define HAVE_GIRL_RECORD
	* src/Makefile.am: Add GIRL_HELPER_PLAYER, GIRL_HELPER_RECORD
	* src/girl-gui.c: Add menu item if GIRL_HELPER_RECORD is defined
	* src/girl-program.c: New file with support for programs
	* src/girl-station.c: Rename girl_launch_helper to girl_helper_run
	* src/girl-station.h: Remove GirlChannels and GirlStreamType
	* src/girl.c: Add programs feature
	* src/girl.h: Add enums GirlChannels, GirlStreamType, GirlHelperType

2014-12-06  Ole Aamot  <oka@oka.no>

	* girl version 0.6.0 release.
	* AUTHORS: Add Wolfgang Stöggl (German translation).
	* po/de.po: German translation by Wolfgang Stöggl

2014-11-29  Ole Aamot  <oka@oka.no>

	* girl version 0.5.2 release.
	* girl.spec.in: Update package spec for Fedora 21.
	* man/girl.1: Add man page.

2014-11-29  Ole Aamot  <oka@oka.no>

	* AUTHORS: Add Мирослав Николић/Miroslav Nikolić (Serbian translation).
	* NEWS: girl 0.5.1 released.
	* po/sr.po: Serbian translation by Мирослав Николић
	* po/sr@latin.po: Serbian translation by Miroslav Nikolić
	* src/girl-station.[ch]: Add station rank, type variables

2014-11-29  Ole Aamot  <oka@oka.no>

	* girl version 0.5.0 release
	* src/girl.c: Initial Previous Station / Next Station support
	* src/girl-gui.c: Add Previous Station / Next Station buttons
	* src/stations.xml: Add Radio Campus Paris (radiocampusparis.org).
	* src/stations.xml: Add WCSB (wcsb.org).
	* src/stations.xml: Add Radio Campus de l’Université Libre de Bruxelles (radiocampus.be).
	* src/stations.xml: Add Coimbra University Radio (ruc.pt).

2014-11-27  Ole Aamot  <oka@oka.no>

	* girl version 0.4.1 release
	* AUTHORS: Add Marek Černocký (Czech translation)
	* AUTHORS: Add Rafael Ferreira (Brazilian Portuguese translation)
	* po/cs.po: Czech translation by Marek Černocký
	* po/pt_BR.po: Brazilian Portuguese translation by Rafael Ferreira
	* src/girl-0.5.dtd: Add station type property.
	* src/stations.xml: Add Cam FM (www.camfm.co.uk).
	* src/stations.xml: Add Oxford Student Radio (oxideradio.co.uk).
	* src/stations.xml: Add Pulse LSE (pulselse.co.uk).
	* src/stations.xml: Add Radio UNAM (www.radiounam.unam.mx).

2014-11-22  Ole Aamot  <oka@oka.no>


src/girl.c
/* $Id: girl.c,v 1.4 2002/10/04 20:20:30 oka Exp $
 *
 * GNOME Internet Radio Locator
 *
 * Copyright (C) 2014  Ole Aamot Software
 *
 * Author: Ole Aamot <oka@oka.no>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
 */

#include <config.h>

#include <gnome.h>
#include <libgnomevfs/gnome-vfs.h>
#include <libgnomevfs/gnome-vfs-application-registry.h>

#include <stdlib.h>
#include <string.h>

#include "girl.h"
#include "girl-gui.h"
#include "girl-listener.h"
#include "girl-station.h"

#ifdef GIRL_DEBUG
#define MSG(x...) g_message(x)
#else
#define MSG(x...)
#endif

GirlData *girl;

GList *girl_listeners;
GList *girl_programs;
GList *girl_stations;
GList *girl_streams;

GtkWidget *girl_app;
GtkWidget *listeners_selector = NULL;
GtkWidget *programs_selector = NULL;
GtkWidget *stations_selector = NULL;
GtkWidget *streams_selector = NULL;

int main(int argc, char *argv[])
{
	gchar *iconname;

#ifdef GETTEXT_PACKAGE
	bindtextdomain(GETTEXT_PACKAGE, GNOMELOCALEDIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain(GETTEXT_PACKAGE);
#endif

	if (gnome_vfs_init() == FALSE) {
		g_warning("Gaaaaaaaah! Gnome-VFS init failed mate !\n");
		exit(1);
	}

	gnome_program_init(PACKAGE, VERSION,
			   LIBGNOMEUI_MODULE,
			   argc, argv,
			   GNOME_PARAM_APP_DATADIR, DATADIR, NULL);

#if defined(G_THREADS_ENABLED) && ! defined(G_THREADS_IMPL_NONE)
	g_thread_init (NULL);
#endif

	girl_app = create_girl_app();

	gtk_widget_show(girl_app);

	/* Set up the listeners list */
	/* listeners_selector = create_listeners_selector(girl->selected_listener_uri,"listeners.xml"); */

	/* g_object_add_weak_pointer(G_OBJECT(listeners_selector), */
	/* 			  (void **) &(listeners_selector)); */

	/* /\* Set up the programs list *\/ */
	/* programs_selector = create_programs_selector(girl->selected_program_uri,"programs.xml"); */

	/* gtk_widget_show(programs_selector); */

        g_object_add_weak_pointer(G_OBJECT(programs_selector),
				  (void **) &(programs_selector));

	/* Set up the stations list */
	stations_selector = create_stations_selector(girl->selected_station_uri,
						     "stations.xml");

	/* Set up the streams list */
	streams_selector = create_streams_selector(girl->selected_streams_uri,
						   "streams.xml");

	g_object_add_weak_pointer(G_OBJECT(stations_selector),
				  (void **) &(stations_selector));

	g_object_add_weak_pointer(G_OBJECT(streams_selector),
				  (void **) &(streams_selector));
	/* Icons */
	iconname = gnome_program_locate_file(NULL,
					     GNOME_FILE_DOMAIN_APP_PIXMAP,
					     "girl/icons/hicolor/48x48/apps/girl.png", TRUE, NULL);
	if (iconname != NULL) {
		girl->icon = gdk_pixbuf_new_from_file(iconname, NULL);
		g_free(iconname);
	} else {
		girl->icon = NULL;
	}

	if (girl->icon != NULL)
		gtk_window_set_icon(GTK_WINDOW(girl_app), girl->icon);

	gtk_main();
	return 0;
}

void
toggle_in_area (double x1, double y1, double x2, double y2)
{
	double nx1,nx2,ny1,ny2;
	
}

void appbar_send_msg(const char *a, ...)
{
	gchar foo[2048];
	va_list ap;
	va_start(ap, a);

	vsnprintf(foo, 2000, a, ap);

	gnome_appbar_pop(girl->appbar);
	gnome_appbar_push(girl->appbar, foo);

	MSG("%s", foo);

	/* lets update the statusbar widget */
	while (gtk_events_pending())
		gtk_main_iteration();

	va_end(ap);
}

static void cause_movement(int way)
{

}

void on_previous_station_click(GtkWidget * a, gpointer user_data)
{
	GList *l = g_list_previous(girl_stations);
	if (l != NULL) {
		GirlStationInfo *station = l->data;
		if (station->prev != NULL) {
			girl_stations->data = station->prev;
		} else {
			girl_stations->data = station;
		}

		cause_movement(-1);
		if (station != NULL) {
			printf("Previous Station ID: %s\n", station->id);
			printf("Previous Station Name: %s\n", station->name);
			printf("Previous Station URI: %s\n", station->stream->uri);
			printf("Previous Station Location: %s\n", station->location);

			girl->selected_station_uri = station->stream->uri;

			MSG("on_station_select_changed: %s\n",
			    girl->selected_station_uri);

			girl->selected_station_name = station->name;

			MSG("on_station_select_changed: %s\n",
			    girl->selected_station_name);

			girl->selected_station_location = station->location;

			MSG("on_station_select_changed: %s\n",
			    girl->selected_station_location);

			girl->selected_station_release = station->release;

			MSG("on_station_select_changed: %s\n",
			    girl->selected_station_release);

			girl->selected_station_description = station->description;

			MSG("on_station_select_changed: %s\n",
			    girl->selected_station_description);

			appbar_send_msg(_("Selected %s in %s: %s"),
					girl->selected_station_name,
					girl->selected_station_location,
					girl->selected_station_uri,
					girl->selected_station_release);
			

			girl_helper_run(girl->selected_station_uri,
					girl->selected_station_name,
					GIRL_STREAM_SHOUTCAST,
					GIRL_STREAM_PLAYER);
		}
	} else {
		printf("Prev Station: At the beginning of Stations list!\n");
	}
}

void on_next_station_click(GtkWidget * a, gpointer user_data)
{
	GList *l = g_list_first(girl_stations);
	if (l != NULL) {
		GirlStationInfo *station = l->data;
		if (station->next != NULL) {
			girl_stations->data = station->next;
		} else {
			printf("Next Station: At the end of Stations list!\n");
		}
		cause_movement(1);
		if (station != NULL) {
			printf("Next Station ID: %s\n", station->id);
			printf("Next Station Name: %s\n", station->name);
			printf("Next Station URI: %s\n", station->stream->uri);
			printf("Next Station Location: %s\n", station->location);

			girl->selected_station_uri = station->stream->uri;

			MSG("on_station_select_changed: %s\n",
			    girl->selected_station_uri);

			girl->selected_station_name = station->name;

			MSG("on_station_select_changed: %s\n",
			    girl->selected_station_name);

			girl->selected_station_location = station->location;

			MSG("on_station_select_changed: %s\n",
			    girl->selected_station_location);

			girl->selected_station_release = station->release;

			MSG("on_station_select_changed: %s\n",
			    girl->selected_station_release);

			girl->selected_station_description = station->description;

			MSG("on_station_select_changed: %s\n",
			    girl->selected_station_description);

			appbar_send_msg(_("Selected %s in %s: %s"),
					girl->selected_station_name,
					girl->selected_station_location,
					girl->selected_station_uri,
					girl->selected_station_release);
			
			girl_helper_run(station->stream->uri,
					station->name,
					GIRL_STREAM_SHOUTCAST,
					GIRL_STREAM_PLAYER);
		}
	}
}

void on_listeners_selector_button_clicked(GtkWidget * a,
					  gpointer user_data)
{
	gtk_widget_show(listeners_selector);
	gdk_window_raise(listeners_selector->window);
}

void on_listeners_selector_changed(GtkWidget * a, gpointer user_data)
{
	if (girl->selected_listener_uri != NULL)
		g_free(girl->selected_listener_uri);

	girl->selected_listener_uri =
	    g_strdup(g_object_get_data(G_OBJECT(a), "listener_uri"));
	MSG("on_listener_select_changed: %s\n",
	    girl->selected_listener_uri);

	girl->selected_listener_name =
	    g_strdup(g_object_get_data(G_OBJECT(a), "listener_name"));
	MSG("on_listener_select_changed: %s\n",
	    girl->selected_listener_name);

	girl->selected_listener_location =
	    g_strdup(g_object_get_data(G_OBJECT(a), "listener_location"));
	MSG("on_listener_select_changed: %s\n",
	    girl->selected_listener_location);

	girl->selected_listener_release =
	    g_strdup(g_object_get_data(G_OBJECT(a), "listener_release"));
	MSG("on_listener_select_changed: %s\n",
	    girl->selected_listener_release);

	girl->selected_listener_description =
	    g_strdup(g_object_get_data(G_OBJECT(a), "listener_description"));
	MSG("on_listener_select_changed: %s\n",
	    girl->selected_listener_description);

	appbar_send_msg(_("Selected %s in %s: %s (%f)"),
			girl->selected_listener_name,
			girl->selected_listener_location,
			girl->selected_listener_uri,
		        girl->selected_listener_release);

	girl_helper_run(girl->selected_listener_uri,
			girl->selected_listener_name,
			GIRL_STREAM_SHOUTCAST,
			GIRL_STREAM_PLAYER);

}

void on_programs_selector_button_clicked(GtkWidget * a, gpointer user_data)
{
	gtk_widget_show(programs_selector);
	gdk_window_raise(programs_selector->window);
}

void on_programs_selector_changed(GtkWidget * a, gpointer user_data)
{
	if (girl->selected_program_uri != NULL)
		g_free(girl->selected_program_uri);

	girl->selected_program_uri =
	    g_strdup(g_object_get_data(G_OBJECT(a), "program_uri"));
	MSG("on_program_select_changed: %s\n", girl->selected_program_uri);

	girl->selected_program_name =
	    g_strdup(g_object_get_data(G_OBJECT(a), "program_name"));
	MSG("on_program_select_changed: %s\n",
	    girl->selected_program_name);

	girl->selected_program_location =
	    g_strdup(g_object_get_data(G_OBJECT(a), "program_location"));
	MSG("on_program_select_changed: %s\n",
	    girl->selected_program_location);

	girl->selected_program_release =
	    g_strdup(g_object_get_data(G_OBJECT(a), "program_release"));
	MSG("on_program_select_changed: %s\n",
	    girl->selected_program_release);

	girl->selected_program_description =
	    g_strdup(g_object_get_data(G_OBJECT(a), "program_description"));
	MSG("on_program_select_changed: %s\n",
	    girl->selected_program_description);

	appbar_send_msg(_("Selected %s in %s: %s"),
			girl->selected_program_name,
			girl->selected_program_location,
			girl->selected_program_uri,
			girl->selected_program_release);

	girl_helper_run(girl->selected_program_uri,
			girl->selected_program_name,
			GIRL_STREAM_SHOUTCAST,
			GIRL_STREAM_PLAYER);
}

void on_stations_selector_button_clicked(GtkWidget * a, gpointer user_data)
{
	gtk_widget_show(stations_selector);
	gdk_window_raise(stations_selector->window);
}

void on_stations_selector_changed(GtkWidget * a, gpointer user_data)
{
	if (girl->selected_station_uri != NULL)
		g_free(girl->selected_station_uri);

	girl->selected_station_uri =
	    g_strdup(g_object_get_data(G_OBJECT(a), "station_uri"));
	MSG("on_station_select_changed: %s\n", girl->selected_station_uri);

	girl->selected_station_name =
	    g_strdup(g_object_get_data(G_OBJECT(a), "station_name"));
	MSG("on_station_select_changed: %s\n",
	    girl->selected_station_name);

	girl->selected_station_location =
	    g_strdup(g_object_get_data(G_OBJECT(a), "station_location"));
	MSG("on_station_select_changed: %s\n",
	    girl->selected_station_location);

	girl->selected_station_release =
	    g_strdup(g_object_get_data(G_OBJECT(a), "station_release"));
	MSG("on_station_select_changed: %s\n",
	    girl->selected_station_release);

	girl->selected_station_description =
	    g_strdup(g_object_get_data(G_OBJECT(a), "station_description"));
	MSG("on_station_select_changed: %s\n",
	    girl->selected_station_description);

	appbar_send_msg(_("Selected %s in %s: %s"),
			girl->selected_station_name,
			girl->selected_station_location,
			girl->selected_station_uri,
			girl->selected_station_release);

	girl_helper_run(girl->selected_station_uri,
			girl->selected_station_name,
			GIRL_STREAM_SHOUTCAST,
		        GIRL_STREAM_PLAYER);
}

void on_streams_selector_button_clicked(GtkWidget * a, gpointer user_data)
{
	gtk_widget_show(streams_selector);
	gdk_window_raise(streams_selector->window);
}

void on_streams_selector_changed(GtkWidget * a, gpointer user_data)
{
	if (girl->selected_streams_uri != NULL)
		g_free(girl->selected_streams_uri);

	girl->selected_streams_uri = g_strdup(g_object_get_data(G_OBJECT(a), "streams_uri"));
	MSG("on_streams_select_changed: %s\n", girl->selected_streams_uri);

	girl->selected_streams_mime = g_strdup(g_object_get_data(G_OBJECT(a), "streams_mime"));
	MSG("on_streams_select_changed: %s\n",
	    girl->selected_streams_mime);

	girl->selected_streams_codec = g_strdup(g_object_get_data(G_OBJECT(a), "streams_codec"));
	MSG("on_streams_select_changed: %s\n",
	    girl->selected_streams_codec);

	girl->selected_streams_samplerate = g_strdup(g_object_get_data(G_OBJECT(a), "streams_samplerate"));
	MSG("on_streams_select_changed: %s\n",
	    girl->selected_streams_samplerate);

	girl->selected_streams_bitrate = g_strdup(g_object_get_data(G_OBJECT(a), "streams_bitrate"));
	MSG("on_streams_select_changed: %s\n",
	    girl->selected_streams_bitrate);

	girl->selected_streams_channels = g_strdup(g_object_get_data(G_OBJECT(a), "streams_channels"));
	MSG("on_streams_select_changed: %s\n",
	    girl->selected_streams_channels);

	appbar_send_msg(_("Selected %s [%s] [%s] [%s]"),
			girl->selected_streams_uri,
			girl->selected_streams_mime,
			girl->selected_streams_codec,
			girl->selected_streams_bitrate);

	girl_helper_run(girl->selected_streams_uri,
			girl->selected_streams_mime,
			GIRL_STREAM_SHOUTCAST,
		        GIRL_STREAM_PLAYER);
}

void quit_app(GtkWidget * a, gpointer user_data)
{
	gnome_config_push_prefix("/girl/General/");
	gnome_config_set_string("selected_listener_uri",
				girl->selected_listener_uri);
	gnome_config_set_string("selected_listener_name",
				girl->selected_listener_name);
	gnome_config_set_string("selected_listener_location",
				girl->selected_listener_location);
	gnome_config_set_string("selected_listener_release",
				girl->selected_listener_release);
	gnome_config_set_string("selected_station_uri",
				girl->selected_station_uri);
	gnome_config_set_string("selected_station_name",
				girl->selected_station_name);
	gnome_config_set_string("selected_station_location",
				girl->selected_station_location);
	gnome_config_set_string("selected_station_release",
				girl->selected_station_release);
	gnome_config_set_string("selected_station_description",
				girl->selected_station_description);
	gnome_config_set_string("selected_streams_uri",
				girl->selected_streams_uri);
	gnome_config_set_string("selected_streams_mime",
				girl->selected_streams_mime);
	gnome_config_set_string("selected_streams_codec",
				girl->selected_streams_codec);
	gnome_config_set_string("selected_streams_samplerate",
				girl->selected_streams_samplerate);
	gnome_config_set_string("selected_streams_bitrate",
				girl->selected_streams_bitrate);
	gnome_config_set_string("selected_streams_channels",
				girl->selected_streams_channels);
	gnome_config_sync();
	gnome_config_pop_prefix();

	gtk_widget_destroy(girl_app);
	if (GTK_IS_WIDGET(stations_selector)) {
		gtk_widget_destroy(stations_selector);
	}
	if (GTK_IS_WIDGET(streams_selector)) {
		gtk_widget_destroy(streams_selector);
	}

	gtk_main_quit();
}

void about_app(GtkWidget * a, gpointer user_data)
{
	static GtkWidget *about = NULL;
	const gchar *authors[] = {
		("Ole Aamot <ole@src.gnome.org>"),
		NULL,
	};
	const gchar *translator_credits = _("translator_credits");

	if (about) {
		gdk_window_raise(about->window);
		return;
	}

	about = gnome_about_new(_("GNOME Internet Radio Locator"), VERSION, _("Copyright 2014 Ole Aamot Software\nCopyright 2002 Free Software Foundation"), _("Internet Radio Locator"), (const gchar **) authors, NULL,	/* documenters */
				strcmp(translator_credits,
				       "translator_credits") !=
				0 ? translator_credits : NULL, girl->icon);

	g_signal_connect(G_OBJECT(about), "destroy",
			 G_CALLBACK(gtk_widget_destroy), NULL);
	g_signal_connect(G_OBJECT(about), "delete-event",
			 G_CALLBACK(gtk_widget_destroy), NULL);
	g_object_add_weak_pointer(G_OBJECT(about), (void **) &(about));
	gtk_widget_show(about);
}

void about_listener(GtkWidget * a, gpointer user_data)
{
	static GtkWidget *about_listener = NULL;
	const gchar *translator_credits = _("translator_credits");
	const gchar *authors[] = {
		girl->selected_listener_name,
		NULL,
	};

	if (about_listener) {
		gdk_window_raise(about_listener->window);
		return;
	}

	about_listener = gnome_about_new(girl->selected_listener_name,
					girl->selected_listener_release,
					girl->selected_listener_location,
					girl->selected_listener_description,
					authors,
					NULL,
					NULL,
					girl->icon);
	g_signal_connect(G_OBJECT(about_listener), "destroy",
			 G_CALLBACK(gtk_widget_destroy), NULL);
	g_signal_connect(G_OBJECT(about_listener), "delete-event",
			 G_CALLBACK(gtk_widget_destroy), NULL);
	g_object_add_weak_pointer(G_OBJECT(about_listener), (void **) &(about_listener));
	gtk_widget_show(about_listener);
}

void about_program(GtkWidget * a, gpointer user_data)
{
	static GtkWidget *about_program = NULL;
	const gchar *translator_credits = _("translator_credits");
	const gchar *authors[] = {
		girl->selected_program_name,
		NULL,
	};

	if (about_program) {
		gdk_window_raise(about_program->window);
		return;
	}

	if (girl->selected_program_name != NULL) {
		about_program = gnome_about_new(girl->selected_program_name,
						girl->selected_program_release,
						girl->selected_program_location,
						girl->selected_program_description,
						authors,
						NULL,
						NULL,
						girl->icon);
		g_signal_connect(G_OBJECT(about_program), "destroy",
				 G_CALLBACK(gtk_widget_destroy), NULL);
		g_signal_connect(G_OBJECT(about_program), "delete-event",
				 G_CALLBACK(gtk_widget_destroy), NULL);
		g_object_add_weak_pointer(G_OBJECT(about_program), (void **) &(about_program));
		gtk_widget_show(about_program);
	}

}

void about_station(GtkWidget * a, gpointer user_data)
{
	static GtkWidget *about_station = NULL;
	const gchar *translator_credits = _("translator_credits");
	const gchar *authors[] = {
		girl->selected_station_name,
		NULL,
	};

	if (about_station) {
		gdk_window_raise(about_station->window);
		return;
	}

	about_station = gnome_about_new(girl->selected_station_name,
					girl->selected_station_release,
					girl->selected_station_location,
					girl->selected_station_description,
					authors,
					NULL,
					NULL,
					girl->icon);

	g_signal_connect(G_OBJECT(about_station), "destroy",
			 G_CALLBACK(gtk_widget_destroy), NULL);
	g_signal_connect(G_OBJECT(about_station), "delete-event",
			 G_CALLBACK(gtk_widget_destroy), NULL);
	g_object_add_weak_pointer(G_OBJECT(about_station), (void **) &(about_station));
	gtk_widget_show(about_station);
}

void about_streams(GtkWidget * a, gpointer user_data)
{
	static GtkWidget *about_streams = NULL;
	const gchar *translator_credits = _("translator_credits");
	const gchar *authors[] = {
		girl->selected_streams_uri,
		NULL,
	};

	if (about_streams) {
		gdk_window_raise(about_streams->window);
		return;
	}

	if (girl->selected_streams_uri != NULL) {
		about_streams = gnome_about_new(girl->selected_streams_uri,
						girl->selected_streams_codec,
						girl->selected_streams_mime,
						girl->selected_streams_bitrate,
						authors,
						NULL,
						NULL,
						girl->icon);
		g_signal_connect(G_OBJECT(about_streams), "destroy",
				 G_CALLBACK(gtk_widget_destroy), NULL);
		g_signal_connect(G_OBJECT(about_streams), "delete-event",
				 G_CALLBACK(gtk_widget_destroy), NULL);
		g_object_add_weak_pointer(G_OBJECT(about_streams), (void **) &(about_streams));
		gtk_widget_show(about_streams);
	}

}

void on_listen_button_clicked(GtkWidget *a, gpointer user_data)
{


	/* girl_stream_player(a, user_data); */

	appbar_send_msg(_("Listening to %s in %s: %s "),
			girl->selected_station_name,
			girl->selected_station_location,
			girl->selected_station_uri,
			girl->selected_station_release);

	girl_helper_run(girl->selected_station_uri,
			girl->selected_station_name,
			GIRL_STREAM_SHOUTCAST,
			GIRL_STREAM_PLAYER);

}

void on_record_button_clicked(GtkWidget *a, gpointer user_data)
{
	GtkWidget *dialog;
	GPid *pid;

	if (girl->selected_station_name != NULL) {
		appbar_send_msg(_("Recording from %s in %s: %s "),
				girl->selected_station_name,
				girl->selected_station_location,
				girl->selected_station_uri,
				girl->selected_station_release);
		girl_helper_run(girl->selected_station_uri,
				girl->selected_station_name,
				GIRL_STREAM_SHOUTCAST,
				GIRL_STREAM_RECORD);
		girl->selected_runners = girl_runners_new(getpid(),
							  girl->selected_station_name,
							  "date",
							  "time",
							  "girl.wav");
	} else {
		dialog = gtk_message_dialog_new(GTK_WINDOW(girl_app),
						GTK_DIALOG_MODAL,
						GTK_MESSAGE_ERROR,
						GTK_BUTTONS_CLOSE, "Could not record!", NULL);
		gtk_dialog_run(GTK_DIALOG(dialog));
		gtk_widget_destroy(dialog);
	}
}

GnomeVFSFileSize get_size(GnomeVFSURI * uri)
{
	GnomeVFSFileInfoOptions info_options;
	GnomeVFSFileInfo *info;
	GnomeVFSFileSize len;
	GnomeVFSResult result;

	info_options = GNOME_VFS_FILE_INFO_DEFAULT;

	info = gnome_vfs_file_info_new();
	result = gnome_vfs_get_file_info_uri(uri, info, info_options);

	if (result != GNOME_VFS_OK)
		return 0;

	len = (unsigned long long) info->size;
	gnome_vfs_file_info_unref(info);

	return len;
}

static gint
girl_archive_progress_callback(GnomeVFSXferProgressInfo * info, gpointer data)
{

	while (gtk_events_pending())
		gtk_main_iteration();

	appbar_send_msg(_("Archiving %s"), (gchar *) data);
	MSG("%s", gnome_vfs_result_to_string(info->status));
	return TRUE;
}

GnomeVFSURI *girl_archive_new(gchar *src, gchar *dest)
{
	GnomeVFSURI *src_uri, *dest_uri;
	GnomeVFSResult result;
	GnomeVFSXferOptions xfer_options;

	xfer_options = GNOME_VFS_XFER_DEFAULT;

	src_uri = gnome_vfs_uri_new(src);
	dest_uri = gnome_vfs_uri_new(dest);

	result = gnome_vfs_xfer_uri(src_uri,
				    dest_uri,
				    xfer_options,
				    GNOME_VFS_XFER_ERROR_MODE_QUERY,
				    GNOME_VFS_XFER_OVERWRITE_MODE_REPLACE,
				    girl_archive_progress_callback, src);

	gnome_vfs_uri_unref(src_uri);

	if (result != GNOME_VFS_OK)
		return NULL;

	return dest_uri;
}
